From b347919f100e8115632e39bc4c3d22a26bc651a2 Mon Sep 17 00:00:00 2001
From: Paul Keith <javelinanddart@gmail.com>
Date: Fri, 13 Jul 2018 01:51:46 +0200
Subject: [PATCH 1/2] Improve compatibility with older devices

* Only activate and deactivate the first 3GPP and 3GPP2 apps
* This is largely based off of the simActivation hack code

Change-Id: I00458e5095ac155168740f2943715b60bda5714b
---
 .../telephony/LineageExtTelephony.java        | 22 ++++++++++++++++++-
 1 file changed, 21 insertions(+), 1 deletion(-)

diff --git a/src/com/qualcomm/qti/internal/telephony/LineageExtTelephony.java b/src/com/qualcomm/qti/internal/telephony/LineageExtTelephony.java
index d4ea474..4538690 100644
--- a/src/com/qualcomm/qti/internal/telephony/LineageExtTelephony.java
+++ b/src/com/qualcomm/qti/internal/telephony/LineageExtTelephony.java
@@ -32,6 +32,7 @@ import android.telephony.TelephonyManager;
 import com.android.internal.telephony.CommandsInterface;
 import com.android.internal.telephony.Phone;
 import com.android.internal.telephony.PhoneConstants;
+import com.android.internal.telephony.uicc.IccCardApplicationStatus.AppType;
 import com.android.internal.telephony.uicc.IccCardStatus.CardState;
 import com.android.internal.telephony.uicc.UiccCard;
 import com.android.internal.telephony.uicc.UiccController;
@@ -49,6 +50,11 @@ import static android.telephony.TelephonyManager.SIM_ACTIVATION_STATE_DEACTIVATE
 
 import static android.telephony.TelephonyManager.MultiSimVariants.DSDA;
 
+import static com.android.internal.telephony.uicc.IccCardApplicationStatus.AppType.APPTYPE_CSIM;
+import static com.android.internal.telephony.uicc.IccCardApplicationStatus.AppType.APPTYPE_RUIM;
+import static com.android.internal.telephony.uicc.IccCardApplicationStatus.AppType.APPTYPE_SIM;
+import static com.android.internal.telephony.uicc.IccCardApplicationStatus.AppType.APPTYPE_USIM;
+
 import static com.android.internal.telephony.uicc.IccCardStatus.CardState.CARDSTATE_PRESENT;
 
 public class LineageExtTelephony extends IExtTelephony.Stub {
@@ -197,6 +203,8 @@ public class LineageExtTelephony extends IExtTelephony.Stub {
         UiccCard card = mPhones[slotId].getUiccCard();
 
         int numApps = card.getNumApplications();
+        int gsmIndex = -1;
+        int cdmaIndex = -1;
 
         mUiccStatus[slotId].mProvisioned = activate;
 
@@ -205,7 +213,19 @@ public class LineageExtTelephony extends IExtTelephony.Stub {
                 continue;
             }
 
-            mCommandsInterfaces[slotId].setUiccSubscription(i, activate, null);
+            AppType appType = card.getApplicationIndex(i).getType();
+            if (gsmIndex < 0 && (appType == APPTYPE_USIM || appType == APPTYPE_SIM)) {
+                gsmIndex = i;
+            } else if (cdmaIndex < 0 && (appType == APPTYPE_CSIM || appType == APPTYPE_RUIM)) {
+                cdmaIndex = i;
+            }
+        }
+
+        if (gsmIndex >= 0) {
+            mCommandsInterfaces[slotId].setUiccSubscription(gsmIndex, activate, null);
+        }
+        if (cdmaIndex >= 0) {
+            mCommandsInterfaces[slotId].setUiccSubscription(cdmaIndex, activate, null);
         }
     }
 
-- 
2.17.1

